### PR TITLE
[WIN32SS][NTUSER][USER32] Make possible to set custom mouse cursors

### PR DESCRIPTION
**Pull request by oleg-dubinskiy**
## Purpose

Allow loading custom mouse cursors set by user from Control Panel.

JIRA issue: [CORE-14165](https://jira.reactos.org/browse/CORE-14165), [CORE-14166](https://jira.reactos.org/browse/CORE-14166).

## Proposed changes

[NTUSER]
- Implement SPI_SETCURSORS case for SystemParametersInfoA/W. According to MSDN, it updates the system mouse cursors by the ones provided by user (can be set via main.cpl). It does not use any parameters from SystemParametersInfo.
1. First, get the cursor path from user defined values in registry via win32k!RegReadUserSetting.
2. Then load the cursor handle from the specified cursor via user32!LoadImageW called from the aprropriate win32k callback via KeUserModeCallback.
3. Set received handle for an appropriate resource ID via win32k!NtUserSetSystemCursor. Do this for each system defined cursor.
- Implement co_IntLoadImage callback for user32!LoadImageW. Add an appropriate part in user32 also.
- Rewrite co_IntSetupDefaultCursors callback, responsible for cursors loading. Make the cursors initialization same as system icons initialization. Do not use NtUserSetSystemCursor for loading, since Windows doesn't do that. Add an internal IntLoadSytemCursors routine instead, equivalently to IntLoadSystemIcons.
- Properly unload system cursors and icons on shutdown in win32k!ExitThreadCallback, when the last thread is terminating.
- Refactor some other several cursor/icon functions: NtUserFindExistingCursorIcon, NtUserSetSystemCursor and DefWndHandleSetCursor.
- Handle HTHELP case in win32k!GetNCHitEx and user32!CreateDialogIndirectA/W, which is responsible for help button class. Set an appropriate cursor for this case in DefWndHandleSetCursor (DefWindowProc!WM_SETCURSOR message).
- Fix initialization of gSharedInfo for kernel mode. It is required by UserValidateHandle* routines.
- Implement UserValidateHandle and UserValidateHandleNoError routines and validate the class cursor via them using helper CheckAndLockCursorIcon fucntion called from IntCreateClass. Don't assign default window class cursor to it, since it is always default as-is, so due to this mouse cursor remains unchaned in the whole client area. What CheckAndLockCursorIcon does, is simply validating passed handle and assigning validation result to it back. Yeah, re-assigning the handle to itself seems to be bogus, but in this case, it's Windows-compatible.
- Add Desktop/SharedPtrToUser routines for validating, equivalently to user mode.
- Remove bogus WM_SETCURSOR handing from win32k!DesktopWindowProc, since it does not load a proper cursor at all, only default IDC_ARROW. It is already handled properly in win32k!IntDefWindowProc.
- Set correct GreSetPointerShape flags for animated mouse cursors.

[USER32]
- Add/fix user mode parts of LoadImage and SetDefaultCursors callbacks. Don't try to load system cursor scheme, it should be done in main.cpl instead.
- Partially implement support for anuimated mouse cursors (*.ani). Load them correcly in CURSORICON_LoadFromFileW. We already have CURSORICON_GetCursorDataFromANI, which handles it properly. Also set the correct flags for CURSORDATA structure and enable CURSORF_ACON flag on cursor creation in case cursor is animated.

[WINSRV]
- Fix error in LoadCursorW call when loading default cursor. IDC_ARROW is already defined as MAKEINTRESOURCE(OCR_ARROW), so placing it into MAKEINTRESOURCE again causes double retrieving the resource. To avoid this, we should use either MAKEINTRESOURCE(OCR_ARROW), or just IDC_ARROW. I user the 2nd option. **TODO**: fix unchanged mouse cursor in a Command Prompt window. Currently, it is still always set to default in a client area of Command Prompt. I suppose that this bug comes from winsrv, since cmd is associated with that, but I don't know what causes this bug exactly. Btw the same bug happens with browseui toolbar. :slightly_frowning_face: 

## TODO

- [ ] Make ani cursors actually animated. They are loaded properly now, but only the first frame is set as cursor. We should load all other frames also.
- [ ] Save user's cursors propely after reboot. Fix saving user's cursors in PROCESSINFO.pCursorCache of PCURICON_OBJECT, to make them available at next boot. Currently, our cursor/icon cache management seems to be broken.

## RESULT

As a proof of the result, see the following video with applying cursors demonstration. :slightly_smiling_face: 

https://user-images.githubusercontent.com/26385117/199766519-a1fd6e9d-566e-4399-9d1a-0dce773af1b6.mp4

### CODE REVIEW

File: win32ss/user/user32/windows/cursoricon.c [hunk 1]
```diff
@@ -1446,6 +1426,7 @@ CURSORICON_LoadImageW(
         /* Prepare the resource name string */
         if(IS_INTRESOURCE(lpszName))
         {
+            ERR("1\n");
```
Comment (binarymaster) [inline win32ss/user/user32/windows/cursoricon.c:1429]: Leftover debug.

**Review by binarymaster — COMMENTED** [review]
I'm not a Win32k expert, but here's something to think about.

File: win32ss/user/ntuser/callback.c [hunk 2]
```diff
@@ -980,6 +1007,61 @@ co_IntClientThreadSetup(VOID)
...
+
+   Argument = IntCbAllocateMemory(ArgumentLength);
+   if (NULL == Argument)
+   {
+      ERR("LoadImage callback failed: out of memory\n");
+      return 0;
+   }
+   Common = (PLOADIMAGE_CALLBACK_ARGUMENTS) Argument;
+   TRACE("Name %ls\n", name);
+   wcscpy(Common->ImageName, name);
```
Comment (katahiromz) [inline win32ss/user/ntuser/callback.c:1033]: `RtlStringCchCopyW`

File: win32ss/user/ntuser/class.c [hunk 3]
```diff
@@ -1037,6 +1037,20 @@ IntCheckProcessDesktopClasses(IN PDESKTOP Desktop,
...
+BOOL
+CheckAndLockCursorIcon(PCURICON_OBJECT *ppcur)
+{
+    PCURICON_OBJECT pcur;
+
+    pcur = UserValidateHandleNoError(*ppcur, TYPE_CURSOR);
+    //if (!pcur)
+        //return FALSE;
+
+    UserAssignmentLock((PVOID)ppcur, pcur);
```
Comment (katahiromz) [inline win32ss/user/ntuser/class.c:1050]:
```suggestion
    UserAssignmentLock((PVOID*)ppcur, pcur);
```
???

File: win32ss/user/ntuser/cursoricon.c [hunk 4]
```diff
@@ -134,52 +139,98 @@ IntRemoveCursorFromList(
...
+
+    for (i = 0 ; i < 16; i++)
     {
-        pcur = UserGetCurIconObject(hcur);
-        if (!pcur)
+        if (gasyscur[i].type == id)
         {
-            EngSetLastError(ERROR_INVALID_CURSOR_HANDLE);
+            pcur->CURSORF_flags |= CURSORF_GLOBAL;
+            UserAssignmentLock((PVOID)&gasyscur[i].handle, pcur);
```
Comment (katahiromz) [inline win32ss/user/ntuser/cursoricon.c:165]:
```suggestion
            UserAssignmentLock((PVOID*)&gasyscur[i].handle, pcur);
```

File: win32ss/user/ntuser/cursoricon.c [hunk 5]
```diff
@@ -134,52 +139,98 @@ IntRemoveCursorFromList(
...
+        gpsi->hIconSmWindows = UserHMGetHandle(pcur);
+        return;
+    }
+
+    for (i = 0 ; i < 6; i++)
+    {
+        if (gasysico[i].type == id)
         {
             pcur->CURSORF_flags |= CURSORF_GLOBAL;
+            UserAssignmentLock((PVOID)&gasysico[i].handle, pcur);
```
Comment (katahiromz) [inline win32ss/user/ntuser/cursoricon.c:205]:
```suggestion
            UserAssignmentLock((PVOID*)&gasysico[i].handle, pcur);
```

File: win32ss/user/ntuser/cursoricon.c [hunk 6]
```diff
@@ -134,52 +139,98 @@ IntRemoveCursorFromList(
...
+    PCURICON_OBJECT pcur;
+    int i;
 
-                //
-                //  The active switch between LR shared and Global public.
-                //  This is hacked around to support this while at the initial system start up.
-                //
-                pcur->head.ppi = NULL;
+    /* Unload icons */
+    for (i = 0; i < 6; i++)
```
Comment (katahiromz) [inline win32ss/user/ntuser/cursoricon.c:219]: magic number 6.

File: win32ss/user/ntuser/cursoricon.c [hunk 7]
```diff
@@ -134,52 +139,98 @@ IntRemoveCursorFromList(
...
+        pcur->head.ppi = PsGetCurrentProcessWin32Process();
+        UserAssignmentUnlock((PVOID)&gasysico[i].handle);
+    }
 
-                IntInsertCursorIntoList(pcur);
-                return;
-            }
-        }
+    /* Unload cursors */
+    for (i = 0; i < 16; i++)
```
Comment (katahiromz) [inline win32ss/user/ntuser/cursoricon.c:228]: magic number 16.

File: win32ss/user/ntuser/cursoricon.c [hunk 8]
```diff
@@ -134,52 +139,98 @@ IntRemoveCursorFromList(
...
+    {
+        pcur = gasyscur[i].handle;
+        if (!pcur) continue;
+        pcur->head.ppi = PsGetCurrentProcessWin32Process();
+        UserAssignmentUnlock((PVOID)&gasyscur[i].handle);
```
Comment (katahiromz) [inline win32ss/user/ntuser/cursoricon.c:233]:
```suggestion
        UserAssignmentUnlock((PVOID*)&gasyscur[i].handle);
```

File: win32ss/user/ntuser/defwnd.c [hunk 9]
```diff
@@ -241,123 +241,150 @@ co_IntFindChildWindowToOwner(PWND Root, PWND Owner)
...
-	 break;
-      }
+            ////
+            if (Msg == WM_LBUTTONDOWN || Msg == WM_MBUTTONDOWN ||
+                Msg == WM_RBUTTONDOWN || Msg == WM_XBUTTONDOWN)
+            {
+                if (pwndPopUP)
+                {
+                    FLASHWINFO fwi =
+                        {sizeof(FLASHWINFO),
```
Comment (katahiromz) [inline win32ss/user/ntuser/defwnd.c:300]: Coding Style

File: win32ss/user/ntuser/callback.c [hunk 10]
```diff
@@ -980,6 +1007,61 @@ co_IntClientThreadSetup(VOID)
...
    return Status;
 }
+ 
+HANDLE FASTCALL
+co_IntLoadImage(LPCWSTR name, UINT type, INT desiredx, INT desiredy, UINT flags)
+{
+   HANDLE Handle;
+   NTSTATUS Status;
+   ULONG ArgumentLength, ResultLength;
+   PVOID Argument, ResultPointer;
```
Comment (peterooch) [inline win32ss/user/ntuser/callback.c:1017]: `Argument` is not needed, you can use `Common` directly in `KeUserModeCallback` without any casting.

Use can see an example here: https://github.com/reactos/reactos/blob/be970d8d313e9af71990c8b6025e8fcfa2f0c1b6/win32ss/user/rtl/text.c#L971

Comment (HBelusca) [reply win32ss/user/ntuser/callback.c:1017]: There are cases where the capture buffer is needed, though.

File: win32ss/user/ntuser/sysparams.c [hunk 11]
```diff
@@ -57,6 +58,23 @@ static const WCHAR* VAL_HOVERWIDTH = L"MouseHoverWidth";
 static const WCHAR* VAL_HOVERHEIGHT = L"MouseHoverHeight";
 static const WCHAR* VAL_SENSITIVITY = L"MouseSensitivity";
 
+static const WCHAR* KEY_CURSORS = L"Control Panel\\Cursors";
```
Comment (katahiromz) [inline win32ss/user/ntuser/sysparams.c:61]: LPCWSTR

File: win32ss/user/ntuser/ntuser.c [hunk 12]
```diff
@@ -199,6 +200,11 @@ NtUserInitialize(
     /* Save the EPROCESS of CSRSS */
     InitCsrProcess(/*PsGetCurrentProcess()*/);
 
+    /* Initialize global shared info */
```
Comment (HBelusca) [inline win32ss/user/ntuser/ntuser.c:203]: This gSharedInfo thing is already being done in
https://github.com/HBelusca/reactos/commit/456d1015ab714d78e4ff0ca280a364e74e26857c

File: win32ss/user/user32/windows/cursoricon.c [hunk 13]
```diff
@@ -18,57 +18,34 @@ WINE_DECLARE_DEBUG_CHANNEL(icon);
...
-        hCursor = LoadCursorW(0, IDC_ARROW);
-        SetCursor(hCursor);
-    }
-
-    return(ZwCallbackReturn(&hCursor, sizeof(HCURSOR), STATUS_SUCCESS));
+    PLOADCURSORS_CALLBACK_ARGUMENTS Common = Arguments;
+
+    ERR("Loading System Cursors\n");
+
+    Common->hCursorArrow       = LoadImageW(NULL, IDC_ARROW,       IMAGE_CURSOR, 0, 0, LR_DEFAULTSIZE | LR_SHARED);
```
Comment (katahiromz) [inline win32ss/user/user32/windows/cursoricon.c:31]: Did you try color cursors?

File: win32ss/user/ntuser/callback.c [hunk 14]
```diff
@@ -465,39 +465,66 @@ co_IntLoadSysMenuTemplate(VOID)
...
 
    /* Do not allow the desktop thread to do callback to user mode */
    ASSERT(PsGetCurrentThreadWin32Thread() != gptiDesktopThread);
 
    ResultPointer = NULL;
-   ResultLength = sizeof(HCURSOR);
+   ResultLength = ArgumentLength = sizeof(LOADCURSORS_CALLBACK_ARGUMENTS);
+
+   Argument = IntCbAllocateMemory(ArgumentLength);
+   if (NULL == Argument)
```
[the remark this answers is not in the document]
Comment (oleg-dubinskiy) [reply win32ss/user/ntuser/callback.c]: Fixed.

File: win32ss/user/ntuser/callback.c [hunk 15]
```diff
@@ -980,6 +1007,61 @@ co_IntClientThreadSetup(VOID)
...
+   PVOID Argument, ResultPointer;
+   PLOADIMAGE_CALLBACK_ARGUMENTS Common;
+
+   ArgumentLength = ResultLength = 0;
+   Argument = ResultPointer = NULL;
+
+   ArgumentLength = sizeof(LOADIMAGE_CALLBACK_ARGUMENTS);
+
+   Argument = IntCbAllocateMemory(ArgumentLength);
+   if (NULL == Argument)
```
[the remark this answers is not in the document]
Comment (oleg-dubinskiy) [reply win32ss/user/ntuser/callback.c]: Fixed.

File: win32ss/user/ntuser/class.c [hunk 16]
```diff
@@ -1037,6 +1037,20 @@ IntCheckProcessDesktopClasses(IN PDESKTOP Desktop,
...
 
+static
+BOOL
+CheckAndLockCursorIcon(PCURICON_OBJECT *ppcur)
+{
+    PCURICON_OBJECT pcur;
+
+    pcur = UserValidateHandleNoError(*ppcur, TYPE_CURSOR);
+    //if (!pcur)
+        //return FALSE;
```
[the remark this answers is not in the document]
Comment (oleg-dubinskiy) [reply win32ss/user/ntuser/class.c]: I disabled it only for testing, to see what will happen if I'll allow to pass invalid cursor handle.
Ideally, it shouldn't be passed in, and I see no any changes with or without this check enabled, so enabled it back.

File: win32ss/user/ntuser/class.c [hunk 17]
```diff
@@ -1948,17 +1960,20 @@ UserSetClassLongPtr(IN PCLS Class,
...
             else
             {
+                ERR("3\n");
                 Ret = 0;
             }
 
             if (Ret == NewLong)
             {
                 /* It's a nop */
+                ERR("4\n");
```
[the remark this answers is not in the document]
Comment (oleg-dubinskiy) [reply win32ss/user/ntuser/class.c]: Yes, right. Removed too.

File: win32ss/user/ntuser/class.c [hunk 18]
```diff
@@ -2759,7 +2764,7 @@ NtUserGetClassInfo(
     // If null instance use client.
     if (!hInstance) hInstance = hModClient;
 
-    TRACE("GetClassInfo(%wZ, %p)\n", &SafeClassName, hInstance);
+    ERR("GetClassInfo(%wZ, %p)\n", &SafeClassName, hInstance);
```
[the remark this answers is not in the document]
Comment (oleg-dubinskiy) [reply win32ss/user/ntuser/class.c]: Because it is always prints the debug string, unlike TRACE.
Now I don't need it anymore, so switched it back.

File: win32ss/user/ntuser/cursoricon.c [hunk 19]
```diff
@@ -896,94 +974,24 @@ NtUserFindExistingCursorIcon(
...
     if (CurIcon)
-        Ret = CurIcon->head.h;
+        Ret = UserHMGetHandle(CurIcon);
     UserLeave();
 
 done:
     if (!IS_INTRESOURCE(ustrRsrcSafe.Buffer))
         ExFreePoolWithTag(ustrRsrcSafe.Buffer, TAG_STRING);
-
+    ERR("Ret %p\n", Ret);
```
[the remark this answers is not in the document]
Comment (oleg-dubinskiy) [reply win32ss/user/ntuser/cursoricon.c]: Yep, reverted too.

File: win32ss/user/user32/windows/cursoricon.c (hunk 1 above)
Comment (oleg-dubinskiy) [reply win32ss/user/user32/windows/cursoricon.c]: Removed.

File: win32ss/user/ntuser/callback.c (hunk 10 above)
Comment (oleg-dubinskiy) [reply win32ss/user/ntuser/callback.c:1017]: OK, so I shouldn't use this suggestion, right?

File: win32ss/user/ntuser/cursoricon.c [hunk 20]
```diff
@@ -134,52 +139,98 @@ IntRemoveCursorFromList(
...
+        EngSetLastError(ERROR_INVALID_CURSOR_HANDLE);
+        return;
+    }
+
+    ppi = PsGetCurrentProcessWin32Process();
+
+    if (!(ppi->W32PF_flags & W32PF_CREATEDWINORDC))
+       return;
+
+    for (i = 0 ; i < 16; i++)
```
Comment (JoachimHenze) [inline win32ss/user/ntuser/cursoricon.c:160]: Again the magic number 16. I do recommend to grep for all occurences.

File: win32ss/user/ntuser/cursoricon.c [hunk 21]
```diff
@@ -2212,41 +2219,43 @@ NtUserSetSystemCursor(
...
+                UserAssignmentLock((PVOID)&gasyscur[i].handle, pcurFrame);
+                UserReferenceObject(pcurFrame);
+            }
+            else
+            {
+                pcur->CURSORF_flags |= CURSORF_GLOBAL;
+                pcur->CURSORF_flags &= ~CURSORF_LINKED;
+                UserAssignmentLock((PVOID)&gasyscur[i].handle, pcur);
+                IntInsertCursorIntoList(pcur);
+		    }
```
Comment (JoachimHenze) [inline win32ss/user/ntuser/cursoricon.c:2252]: wrong indentation here of the closing curly brace.

File: win32ss/user/ntuser/defwnd.c [hunk 22]
```diff
@@ -241,123 +241,149 @@ co_IntFindChildWindowToOwner(PWND Root, PWND Owner)
...
-      case HTERROR:
-      {
-         //// This is the real fix for CORE-6129! This was a "Code hole".
-         USER_REFERENCE_ENTRY Ref;
+    switch((short)LOWORD(lParam))
+    {
+        case HTNOWHERE:
+        case HTERROR:
+        {
+            //// This is the real fix for CORE-6129! This was a "Code hole".
```
Comment (JoachimHenze) [inline win32ss/user/ntuser/defwnd.c:265]: why four slashes instead of 2?

File: win32ss/user/ntuser/ntuser.c (hunk 12 above)
Comment (oleg-dubinskiy) [reply win32ss/user/ntuser/ntuser.c:203]: Awww, indeed! :sweat_smile: Didn't saw your branch earlier @HBelusca. But the problem is, this is not in master yet, although my handle validating implementations are depending on this. So I can't just remove it. And I don't know  when you will decide to finish it and merge in master.

File: win32ss/user/ntuser/defwnd.c [hunk 23]
```diff
@@ -241,123 +241,149 @@ co_IntFindChildWindowToOwner(PWND Root, PWND Owner)
...
+                        co_IntSetForegroundWindow(pwndPopUP); // HACK
+                        UserDerefObjectCo(pwndPopUP);
 
-                  // If the change was made, break out.
-                  if (pwndOrigActive != gpqForeground->spwndActive)
-                     break;
-               }
+                        // If the change was made, break out.
+                        if (pwndOrigActive != gpqForeground->spwndActive)
+                        break;
```
Comment (JoachimHenze) [inline win32ss/user/ntuser/defwnd.c:289]: wrong indentation of the break

File: win32ss/user/ntuser/callback.c (hunk 2 above)
Comment (oleg-dubinskiy) [reply win32ss/user/ntuser/callback.c]: Added, thanks.

File: win32ss/user/ntuser/defwnd.c [hunk 24]
```diff
@@ -241,123 +241,149 @@ co_IntFindChildWindowToOwner(PWND Root, PWND Owner)
...
-                     (gpsi->dtCaretBlink >> 3)};
-
-                 // Now shake that window!
-                 IntFlashWindowEx(pwndPopUP, &fwi);
-             }
-	     UserPostMessage(hwndSAS, WM_LOGONNOTIFY, LN_MESSAGE_BEEP, 0);
-	 }
-	 break;
-      }
+            ////
```
Comment (JoachimHenze) [inline win32ss/user/ntuser/defwnd.c:293]: strip those 4 slashes

**Review by JoachimHenze — CHANGES_REQUESTED** [review]

File: win32ss/user/ntuser/class.c (hunk 3 above)
Comment (oleg-dubinskiy) [reply win32ss/user/ntuser/class.c]: Fixed.

File: win32ss/user/ntuser/cursoricon.c (hunk 4 above)
Comment (oleg-dubinskiy) [reply win32ss/user/ntuser/cursoricon.c]: Also fixed.

File: win32ss/user/ntuser/cursoricon.c (hunk 5 above)
Comment (oleg-dubinskiy) [reply win32ss/user/ntuser/cursoricon.c]: This too, thank you.

File: win32ss/user/ntuser/cursoricon.c (hunk 6 above)
Comment (oleg-dubinskiy) [reply win32ss/user/ntuser/cursoricon.c:219]: Yes, this is an amount of system-defined icons. Should I change something here? @katahiromz

File: win32ss/user/ntuser/defwnd.c (hunk 9 above)
Comment (oleg-dubinskiy) [reply win32ss/user/ntuser/defwnd.c]: Improved coding style.

File: win32ss/user/ntuser/message.c [hunk 25]
```diff
@@ -1479,7 +1479,7 @@ co_IntSendMessageTimeoutSingle( HWND hWnd,
 
     if (!(Window = UserGetWindowObject(hWnd)))
     {
-        TRACE("SendMessageTimeoutSingle: Invalid handle 0x%p!\n",hWnd);
+        ERR("SendMessageTimeoutSingle: Invalid handle 0x%p!\n",hWnd);
```
Comment (JoachimHenze) [inline win32ss/user/ntuser/message.c:1482]: since you touched the line anyway, I would recommend to strip the final exclamation mark also.

File: win32ss/user/ntuser/sysparams.c (hunk 11 above)
Comment (oleg-dubinskiy) [reply win32ss/user/ntuser/sysparams.c]: Added.

File: win32ss/user/ntuser/msgqueue.c [hunk 26]
```diff
@@ -680,7 +679,8 @@ co_MsqInsertMouseMessage(MSG* Msg, DWORD flags, ULONG_PTR dwExtraInfo, BOOL Hook
                                        MessageQueue->CursorObject->yHotspot,
                                        gpsi->ptCursor.x,
                                        gpsi->ptCursor.y,
-                                       MessageQueue->CursorObject->hbmAlpha ? SPS_ALPHA : 0);
+                                       (MessageQueue->CursorObject->CURSORF_flags & CURSORF_ACON) ?
```
Comment (JoachimHenze) [inline win32ss/user/ntuser/msgqueue.c:682]: undesired EOL-whitespace

File: win32ss/user/user32/windows/cursoricon.c (hunk 13 above)
Comment (oleg-dubinskiy) [reply win32ss/user/user32/windows/cursoricon.c:31]: Yes, all cursor schemes I tested are coloured, not monochrome.

Comment (oleg-dubinskiy) [reply win32ss/user/user32/windows/cursoricon.c:31]: Why did you ask this? :thinking:

File: win32ss/user/ntuser/nonclient.c [hunk 27]
```diff
@@ -2075,6 +2075,14 @@ GetNCHitEx(PWND pWnd, POINT pt)
...
                 }
+
+                /* Check help button */
+                if (ExStyle & WS_EX_CONTEXTHELP && !(ExStyle & WS_EX_TOOLWINDOW))
+                {
+                    rcWindow.right -= UserGetSystemMetrics(SM_CXSIZE);
+                    if (pt.x > rcWindow.right) return HTHELP;
+                }
+
             }
```
Comment (JoachimHenze) [inline win32ss/user/ntuser/nonclient.c:2086]: unintended empty-line.

File: win32ss/user/ntuser/ntuser.c [hunk 28]
```diff
@@ -199,6 +200,11 @@ NtUserInitialize(
...
+    gSharedInfo.psi = gpsi;
+    gSharedInfo.aheList = gHandleTable;
+    gSharedInfo.pDispInfo = NULL; // gpDispInfo
```
Comment (JoachimHenze) [inline win32ss/user/ntuser/ntuser.c:206]: what exactly is the added comment trying to tell the reader of that code?

File: win32ss/user/ntuser/object.c [hunk 29]
```diff
@@ -836,6 +837,157 @@ NtUserValidateHandleSecure(
...
+    // Must have an entry and must be the same type!
+    if ((!Entry) || (Entry->type != Type) || !Entry->ptr)
+        return NULL;
+
+    if (g_ObjectHeapTypeShared[Type])
+        ret = SharedPtrToUser(Entry->ptr);
+    else
+        ret = DesktopPtrToUser(Entry->ptr);
+
+
```
Comment (JoachimHenze) [inline win32ss/user/ntuser/object.c:987]: one empty line should be stripped here

File: win32ss/user/ntuser/sysparams.c [hunk 30]
```diff
@@ -15,6 +15,7 @@
 DBG_DEFAULT_CHANNEL(UserSysparams);
 
 SPIVALUES gspv;
+
```
Comment (JoachimHenze) [inline win32ss/user/ntuser/sysparams.c:18]: that added line is an unnecessary change imho.

File: win32ss/user/user32/windows/cursoricon.c [hunk 31]
```diff
@@ -541,7 +518,7 @@ get_best_icon_file_entry(
     HeapFree(GetProcessHeap(), 0, fakeDir);
     if(i == 0)
     {
-        WARN("Unable to get a fit entry index.\n");
+        ERR("Unable to get a fit entry index.\n");
```
Comment (JoachimHenze) [inline win32ss/user/user32/windows/cursoricon.c:521]: since you touched the line anyway already, omit the trailing dot also.

File: win32ss/user/user32/windows/cursoricon.c [hunk 32]
```diff
@@ -1493,7 +1473,7 @@ CURSORICON_LoadImageW(
     {
         FINDEXISTINGCURICONPARAM param;
 
-        TRACE("Checking for an LR_SHARED cursor/icon.\n");
+        TRACE("Checking for an LR_SHARED cursor/icon ustrModule %wZ, ustrRsrc %wZ.\n", &ustrModule, &ustrRsrc);
```
Comment (JoachimHenze) [inline win32ss/user/user32/windows/cursoricon.c:1476]: since you touched the line anyway, remove the trailing dot also

File: win32ss/user/ntuser/cursoricon.c (hunk 6 above)
Comment (JoachimHenze) [reply win32ss/user/ntuser/cursoricon.c:219]: You could use a define for that and define it at the beginning of the file. Especially if this magic number would appear only within this file. Otherwise (when shared across multiple files), some header should be used for that instead.

File: win32ss/user/ntuser/cursoricon.c (hunk 7 above)
Comment (JoachimHenze) [reply win32ss/user/ntuser/cursoricon.c:228]: Same answer as above. But different to above where you answered that question, we don't know what the 16 means here.

File: win32ss/user/ntuser/desktop.c [hunk 33]
```diff
@@ -1516,7 +1496,7 @@ DesktopWindowProc(PWND Wnd, UINT Msg, WPARAM wParam, LPARAM lParam, LRESULT *lRe
         }
         default:
             TRACE("DWP calling IDWP Msg %d\n",Msg);
-            //*lResult = IntDefWindowProc(Wnd, Msg, wParam, lParam, FALSE);
+            *lResult = IntDefWindowProc(Wnd, Msg, wParam, lParam, FALSE);
```
Comment (yagoulas) [inline win32ss/user/ntuser/desktop.c:1499]: I am not sure if it is appropriate to enable this. This window proc runs in the desktop thread which is a weird environment in general.

File: win32ss/include/callback.h [hunk 34]
```diff
@@ -102,6 +102,15 @@ typedef struct _EVENTPROC_CALLBACK_ARGUMENTS
...
   ULONG_PTR offPfn;
 } EVENTPROC_CALLBACK_ARGUMENTS, *PEVENTPROC_CALLBACK_ARGUMENTS;
+ 
+typedef struct _LOADIMAGE_CALLBACK_ARGUMENTS
+{
+    WCHAR ImageName[MAX_PATH];
+    UINT ImageType;
+    int cxDesired;
+    int cyDesired;
+    UINT fuFlags;
```
Comment (yagoulas) [inline win32ss/include/callback.h:112]: perhaps ImageName can be moved at the end of this struct and make it 0 sized. When allocating this struct you can add the size of the string so as to save some space in the stack.

**Comment by oleg-dubinskiy** [comment]
Closing this PR, since it's too outdated, has a lot of conflicts and is superseded by #7064 now.
